### PR TITLE
Add notebooks to docs

### PR DESCRIPTION
**Pull request by kjun9**
This is roughly working but the output is quite messy: https://stellargraph--1279.org.readthedocs.build/en/1279/demos/index.html

The script currently just takes the directory names containing demo notebooks and uses that as section headings, which is not ideal.

Notebooks also contain markdown headings which are inconsistently styled at the moment, which messes up the TOC tree.

Follow-ups:
* messy sections https://github.com/stellargraph/stellargraph/issues/1294
* binder/colab buttons https://github.com/stellargraph/stellargraph/issues/1293
* broken links https://github.com/stellargraph/stellargraph/issues/1292
* readme as indices https://github.com/stellargraph/stellargraph/issues/1298
* validate script in CI https://github.com/stellargraph/stellargraph/issues/1299

#1159

### CODE REVIEW

File: docs/demos/basics/index.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+basics
+====================================
+
+.. toctree::
```
Comment (timpitman) [inline docs/demos/basics/index.txt:4]: Is there a way to capitalise the contents headings to match the rest of the docs? (possibly just re-name the folders, unless there's a sphinx way)

File: scripts/create-nbsphinx-links.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+  # recurse through parent directories and create index.txt
+  nb_link_dir="$(dirname "$nb_link")"
+  while [ "$nb_link_dir" != "docs" ]; do
+    nb_link_parent_index="$nb_link_dir/index.txt"
+    if [[ ! -f "$nb_link_parent_index" ]]; then
+      echo "Creating '$nb_link_parent_index'"
+      mkdir -p "$nb_link_dir"
+      cat > "$nb_link_parent_index" << EOF
+$(basename "$nb_link_dir")
```
Comment (timpitman) [inline scripts/create-nbsphinx-links.sh:17]: Maybe just this can be capitalised? (ref my comment earlier in the PR?)

Comment (kjun9) [reply scripts/create-nbsphinx-links.sh:17]: Hmm yeah that's a reasonable idea! Seems like every directory is kebab case, except for `community_detection` - I guess I could put a formatting rule in here to replace `-`/`_` with space, and capitalise each word?

Comment (timpitman) [reply scripts/create-nbsphinx-links.sh:17]: I like the concept, but figuring out the sed is tricky!  Possibly awk more suited?

Comment (kjun9) [reply scripts/create-nbsphinx-links.sh:17]: BSD sed doesn't support `\U` to capitalise, so I'm trying something with `awk` now..

File: scripts/create-nbsphinx-links.sh [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      mkdir -p "$nb_link_dir"
+      cat > "$nb_link_parent_index" << EOF
+$section_title
+====================================
+
+.. toctree::
+    :titlesonly:
+    :glob:
+
+    */index
```
Comment (huonw) [inline scripts/create-nbsphinx-links.sh:29]: Why are both of these lines needed? Doesn't the `./*` glob include any `index` files?

Comment (kjun9) [reply scripts/create-nbsphinx-links.sh:29]: Ah yep seems like it 👍 a bit confused then how sphinx resolves the toc tree if it adds all the files from subdirectories in the glob recursively.. but the output seems correct at least

**Comment by kjun9** [comment]
> I think there's two more follow up issues we could have:
> 
> * reusing the README.mds as the indices somehow (one approach would be to convert them into `README.rst`s and just have the `toctree` within them)
> * validating that the committed directory structure matches the one computed with the script on CI so there's no divergence? (Long term I guess it would better for it to be a Python script instead of a bash one, since that's more controlled/flexible.)

Thanks, that makes sense. I've created two more issues #1298 and #1299

**Pull request merged**